In [191]:
import pandas as pd

In [192]:
import numpy as np

In [193]:
import warnings

In [194]:
warnings.filterwarnings('ignore')

In [195]:
book=pd.read_csv(r'C:\Users\deyar\OneDrive\Desktop\BookRecommendationSystem\notebooks\data\Books.csv')
user=pd.read_csv(r'C:\Users\deyar\OneDrive\Desktop\BookRecommendationSystem\notebooks\data\Users.csv')
rating=pd.read_csv(r'C:\Users\deyar\OneDrive\Desktop\BookRecommendationSystem\notebooks\data\Ratings.csv')

In [196]:
book.head(2)

ISBN           Book-Title           Book-Author Year-Of-Publication  \
0  0195153448  Classical Mythology    Mark P. O. Morford                2002   
1  0002005018         Clara Callan  Richard Bruce Wright                2001   

                 Publisher                                        Image-URL-S  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...

In [197]:
user.head(2)

User-ID                   Location   Age
0        1         nyc, new york, usa   NaN
1        2  stockton, california, usa  18.0

In [198]:
rating.head(2)

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5

In [199]:
book_and_rating=book.merge(rating,on='ISBN')

In [200]:
book_and_rating.shape

(1031136, 10)

In [201]:
book_and_rating.head(2)

ISBN           Book-Title           Book-Author Year-Of-Publication  \
0  0195153448  Classical Mythology    Mark P. O. Morford                2002   
1  0002005018         Clara Callan  Richard Bruce Wright                2001   

                 Publisher                                        Image-URL-S  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-L  User-ID  Book-Rating  
0  http://images.amazon.com/images/P/0195153448.0...        2            0  
1  http://images.amazon.com/images/P/0002005018.0...        8            5

In [202]:
book_and_rating.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            2
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            4
User-ID                0
Book-Rating            0
dtype: int64

In [203]:
book_and_rating.duplicated().sum()

0

In [204]:
book_and_rating.dropna(inplace=True)

In [205]:
book_and_rating.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            0
Year-Of-Publication    0
Publisher              0
Image-URL-S            0
Image-URL-M            0
Image-URL-L            0
User-ID                0
Book-Rating            0
dtype: int64

In [206]:
book_and_rating.shape

(1031128, 10)

In [207]:
book_and_rating.columns

Index(['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher',
       'Image-URL-S', 'Image-URL-M', 'Image-URL-L', 'User-ID', 'Book-Rating'],
      dtype='object')

In [208]:
count=book_and_rating.groupby('Book-Title')['Book-Rating'].count().reset_index()

In [209]:
count.columns

Index(['Book-Title', 'Book-Rating'], dtype='object')

In [210]:
count=count.rename(columns={'Book-Rating':'rating_count'})

In [211]:
count.head(2)

Book-Title  rating_count
0   A Light in the Storm: The Civil War Diary of ...             4
1                              Always Have Popsicles             1

In [212]:
avg=book_and_rating.groupby('Book-Title')['Book-Rating'].mean().reset_index()

In [213]:
avg.head()

Book-Title  Book-Rating
0   A Light in the Storm: The Civil War Diary of ...         2.25
1                              Always Have Popsicles         0.00
2               Apple Magic (The Collector's series)         0.00
3   Ask Lily (Young Women of Faith: Lily Series, ...         8.00
4   Beyond IBM: Leadership Marketing and Finance ...         0.00

In [214]:
avg=avg.rename(columns={'Book-Rating':'rating_avg'})
avg.head(2)

Book-Title  rating_avg
0   A Light in the Storm: The Civil War Diary of ...        2.25
1                              Always Have Popsicles        0.00

In [215]:
count.head(2)

Book-Title  rating_count
0   A Light in the Storm: The Civil War Diary of ...             4
1                              Always Have Popsicles             1

In [216]:
count_avg=count.merge(avg,on='Book-Title')

In [217]:
count_avg.head()

Book-Title  rating_count  rating_avg
0   A Light in the Storm: The Civil War Diary of ...             4        2.25
1                              Always Have Popsicles             1        0.00
2               Apple Magic (The Collector's series)             1        0.00
3   Ask Lily (Young Women of Faith: Lily Series, ...             1        8.00
4   Beyond IBM: Leadership Marketing and Finance ...             1        0.00

In [218]:
popular_book=count_avg[count_avg['rating_count']>250]

In [219]:
popular_book.head(2)

Book-Title  rating_count  rating_avg
764                 1984           284    4.454225
818  1st to Die: A Novel           509    3.575639

In [220]:
popular_book.shape

(186, 3)

In [221]:
popular_book=popular_book.sort_values(by='rating_count',ascending=False).head(50)

In [222]:
popular_book.shape

(50, 3)

In [223]:
popular_book.head()

Book-Title  rating_count  rating_avg
234945                 Wild Animus          2502    1.019584
196321   The Lovely Bones: A Novel          1295    4.468726
183568           The Da Vinci Code           898    4.642539
5303               A Painted House           838    3.231504
199232  The Nanny Diaries: A Novel           828    3.530193

In [224]:
book.head(2)

ISBN           Book-Title           Book-Author Year-Of-Publication  \
0  0195153448  Classical Mythology    Mark P. O. Morford                2002   
1  0002005018         Clara Callan  Richard Bruce Wright                2001   

                 Publisher                                        Image-URL-S  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...

In [225]:
popular_book50=popular_book.merge(book,on='Book-Title')

In [226]:
popular_book50.shape

(216, 10)

In [227]:
final50=popular_book50.drop_duplicates('Book-Title')

In [228]:
final50.head()

Book-Title  rating_count  rating_avg        ISBN  \
0                  Wild Animus          2502    1.019584  0971880107   
1    The Lovely Bones: A Novel          1295    4.468726  0316666343   
2            The Da Vinci Code           898    4.642539  0385504209   
8              A Painted House           838    3.231504  044023722X   
13  The Nanny Diaries: A Novel           828    3.530193  0312278586   

        Book-Author Year-Of-Publication                Publisher  \
0      Rich Shapero                2004                  Too Far   
1      Alice Sebold                2002            Little, Brown   
2         Dan Brown                2003                Doubleday   
8      John Grisham                2001  Dell Publishing Company   
13  Emma McLaughlin                2002       St. Martin's Press   

                                          Image-URL-S  \
0   http://images.amazon.com/images/P/0971880107.0...   
1   http://images.amazon.com/images/P/0316666343.0...   
2   http://images.amazon.com/images/P/0385504209.0...   
8   http://images.amazon.com/images/P/044023722X.0...   
13  http://images.amazon.com/images/P/0312278586.0...   

                                          Image-URL-M  \
0   http://images.amazon.com/images/P/0971880107.0...   
1   http://images.amazon.com/images/P/0316666343.0...   
2   http://images.amazon.com/images/P/0385504209.0...   
8   http://images.amazon.com/images/P/044023722X.0...   
13  http://images.amazon.com/images/P/0312278586.0...   

                                          Image-URL-L  
0   http://images.amazon.com/images/P/0971880107.0...  
1   http://images.amazon.com/images/P/0316666343.0...  
2   http://images.amazon.com/images/P/0385504209.0...  
8   http://images.amazon.com/images/P/044023722X.0...  
13  http://images.amazon.com/images/P/0312278586.0...

In [229]:
final50.shape

(50, 10)

In [230]:
final50.columns

Index(['Book-Title', 'rating_count', 'rating_avg', 'ISBN', 'Book-Author',
       'Year-Of-Publication', 'Publisher', 'Image-URL-S', 'Image-URL-M',
       'Image-URL-L'],
      dtype='object')

In [231]:
final50=final50[['Book-Title','Image-URL-M','Book-Author','rating_count','rating_avg']]
final50.head(3)

Book-Title  \
0                Wild Animus   
1  The Lovely Bones: A Novel   
2          The Da Vinci Code   

                                         Image-URL-M   Book-Author  \
0  http://images.amazon.com/images/P/0971880107.0...  Rich Shapero   
1  http://images.amazon.com/images/P/0316666343.0...  Alice Sebold   
2  http://images.amazon.com/images/P/0385504209.0...     Dan Brown   

   rating_count  rating_avg  
0          2502    1.019584  
1          1295    4.468726  
2           898    4.642539

In [232]:
book.head(2)

ISBN           Book-Title           Book-Author Year-Of-Publication  \
0  0195153448  Classical Mythology    Mark P. O. Morford                2002   
1  0002005018         Clara Callan  Richard Bruce Wright                2001   

                 Publisher                                        Image-URL-S  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...

In [233]:
user.head(2)

User-ID                   Location   Age
0        1         nyc, new york, usa   NaN
1        2  stockton, california, usa  18.0

In [234]:
rating.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [235]:
book_and_rating.head(2)

ISBN           Book-Title           Book-Author Year-Of-Publication  \
0  0195153448  Classical Mythology    Mark P. O. Morford                2002   
1  0002005018         Clara Callan  Richard Bruce Wright                2001   

                 Publisher                                        Image-URL-S  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-L  User-ID  Book-Rating  
0  http://images.amazon.com/images/P/0195153448.0...        2            0  
1  http://images.amazon.com/images/P/0002005018.0...        8            5

In [236]:
book_and_rating.shape

(1031128, 10)

In [237]:
ratingatleast200=book_and_rating.groupby('User-ID')['Book-Rating'].count().reset_index()

In [238]:
ratingatleast200.rename(columns={'Book-Rating':'Book-Rating_count'},inplace=True)

In [239]:
ratingatleast200.head()

User-ID  Book-Rating_count
0        2                  1
1        8                 17
2        9                  3
3       10                  1
4       12                  1

In [240]:
atleast200ratinguser=ratingatleast200[ratingatleast200['Book-Rating_count']>=200]

In [241]:
atleast200ratinguser.head()

User-ID  Book-Rating_count
87        254                300
698      2276                456
862      2766                269
919      2977                227
1033     3363                890

In [242]:
atleast200ratinguser.shape

(816, 2)

In [243]:
x=pd.unique(atleast200ratinguser['User-ID'])

In [244]:
filter_data=book_and_rating[book_and_rating['User-ID'].isin(x)]

In [245]:
filter_data.shape

(475002, 10)

In [246]:
filter_data.head(2)

ISBN    Book-Title           Book-Author Year-Of-Publication  \
3  0002005018  Clara Callan  Richard Bruce Wright                2001   
6  0002005018  Clara Callan  Richard Bruce Wright                2001   

               Publisher                                        Image-URL-S  \
3  HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   
6  HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-M  \
3  http://images.amazon.com/images/P/0002005018.0...   
6  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-L  User-ID  Book-Rating  
3  http://images.amazon.com/images/P/0002005018.0...    11676            8  
6  http://images.amazon.com/images/P/0002005018.0...    85526            0

In [247]:
c1=filter_data.groupby('Book-Title')['Book-Rating'].count().reset_index()

In [248]:
c1.rename(columns={'Book-Rating':'Book-Rating_count'},inplace=True)

In [249]:
c1.head()

Book-Title  Book-Rating_count
0   A Light in the Storm: The Civil War Diary of ...                  2
1                              Always Have Popsicles                  1
2               Apple Magic (The Collector's series)                  1
3   Beyond IBM: Leadership Marketing and Finance ...                  1
4   Clifford Visita El Hospital (Clifford El Gran...                  1

In [250]:
atlest50_rating=c1[c1['Book-Rating_count']>=50]

In [251]:
atlest50_rating.shape

(707, 2)

In [252]:
atlest50_rating.head(2)

Book-Title  Book-Rating_count
484                 1984                 71
518  1st to Die: A Novel                160

In [253]:
v=pd.unique(atlest50_rating['Book-Title'])

In [254]:
filter_data.shape

(475002, 10)

In [255]:
final_data2=filter_data[filter_data['Book-Title'].isin(v)]

In [256]:
final_data2.shape

(58823, 10)

In [257]:
final_data2.head(2)

ISBN              Book-Title Book-Author Year-Of-Publication  \
31  0399135782  The Kitchen God's Wife     Amy Tan                1991   
33  0399135782  The Kitchen God's Wife     Amy Tan                1991   

           Publisher                                        Image-URL-S  \
31  Putnam Pub Group  http://images.amazon.com/images/P/0399135782.0...   
33  Putnam Pub Group  http://images.amazon.com/images/P/0399135782.0...   

                                          Image-URL-M  \
31  http://images.amazon.com/images/P/0399135782.0...   
33  http://images.amazon.com/images/P/0399135782.0...   

                                          Image-URL-L  User-ID  Book-Rating  
31  http://images.amazon.com/images/P/0399135782.0...    11676            9  
33  http://images.amazon.com/images/P/0399135782.0...    36836            0

In [258]:
data=final_data2.pivot_table(index='Book-Title',columns='User-ID',values='Book-Rating')

In [259]:
data.head()

User-ID              254     2276    2766    2977    3363    4017    4385    \
Book-Title                                                                    
1984                    9.0     NaN     NaN     NaN     NaN     NaN     NaN   
1st to Die: A Novel     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2nd Chance              NaN    10.0     NaN     NaN     NaN     NaN     NaN   
4 Blondes               NaN     NaN     NaN     NaN     NaN     NaN     NaN   
A Bend in the Road      0.0     NaN     7.0     NaN     NaN     NaN     NaN   

User-ID              6251    6323    6543    ...  271705  273979  274004  \
Book-Title                                   ...                           
1984                    NaN     NaN     NaN  ...    10.0     NaN     NaN   
1st to Die: A Novel     NaN     NaN     9.0  ...     NaN     NaN     NaN   
2nd Chance              NaN     NaN     0.0  ...     NaN     NaN     NaN   
4 Blondes               0.0     NaN     NaN  ...     NaN     NaN     NaN   
A Bend in the Road      NaN     NaN     NaN  ...     NaN     0.0     NaN   

User-ID              274061  274301  274308  275970  277427  277639  278418  
Book-Title                                                                   
1984                    NaN     NaN     NaN     0.0     NaN     NaN     NaN  
1st to Die: A Novel     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
2nd Chance              NaN     NaN     0.0     NaN     NaN     0.0     NaN  
4 Blondes               NaN     NaN     NaN     NaN     NaN     NaN     NaN  
A Bend in the Road      NaN     NaN     NaN     NaN     NaN     NaN     NaN  

[5 rows x 815 columns]

In [260]:
data.shape

(707, 815)

In [261]:
data.fillna(0,inplace=True)

In [294]:
data

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   0.0  ...    10.0   
1st to Die: A Novel                                    9.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
A Bend in the Road                                     0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

In [262]:
b=pd.read_csv(r'C:\Users\deyar\OneDrive\Desktop\BookRecommendationSystem\artifacts\data.csv')

In [263]:
b

254  2276  2766  2977  3363  4017  4385  6251  6323  6543  ...  271705  \
0    9.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    10.0   
1    0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   9.0  ...     0.0   
2    0.0  10.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...     0.0   
3    0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...     0.0   
4    0.0   0.0   7.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...     0.0   
..   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...     ...   
702  0.0   0.0   0.0   7.0   0.0   0.0   0.0   0.0   0.0   0.0  ...     0.0   
703  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...     0.0   
704  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...     0.0   
705  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...     0.0   
706  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...     0.0   

     273979  274004  274061  274301  274308  275970  277427  277639  278418  
0       0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
1       0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
2       0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
3       0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
4       0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
..      ...     ...     ...     ...     ...     ...     ...     ...     ...  
702     9.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
703     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
704     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
705     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
706     0.0     0.0     0.0     8.0     0.0     0.0     0.0     0.0     0.0  

[707 rows x 815 columns]

In [264]:
data.fillna(0,inplace=True)

In [265]:
from sklearn.metrics.pairwise import cosine_similarity

In [266]:
similarity=cosine_similarity(data)

In [267]:
similarity

array([[1.        , 0.0999137 , 0.01189468, ..., 0.11799012, 0.07158663,
        0.04205081],
       [0.0999137 , 1.        , 0.2364573 , ..., 0.07446129, 0.16773875,
        0.14263397],
       [0.01189468, 0.2364573 , 1.        , ..., 0.04558758, 0.04938579,
        0.10796119],
       ...,
       [0.11799012, 0.07446129, 0.04558758, ..., 1.        , 0.07085128,
        0.0196177 ],
       [0.07158663, 0.16773875, 0.04938579, ..., 0.07085128, 1.        ,
        0.10602962],
       [0.04205081, 0.14263397, 0.10796119, ..., 0.0196177 , 0.10602962,
        1.        ]])

In [268]:
similarity[0]

array([1.        , 0.0999137 , 0.01189468, 0.        , 0.05229234,
       0.02703559, 0.08005247, 0.133798  , 0.03177829, 0.03573298,
       0.0226271 , 0.06592522, 0.020304  , 0.09425025, 0.1304464 ,
       0.08089641, 0.10866788, 0.0496928 , 0.02453052, 0.11405415,
       0.        , 0.13965274, 0.07645776, 0.05992325, 0.08499677,
       0.        , 0.06829594, 0.13307519, 0.07404925, 0.11854319,
       0.007483  , 0.01435345, 0.        , 0.07761015, 0.04406341,
       0.0151626 , 0.09251799, 0.01776199, 0.02543351, 0.07778745,
       0.11379679, 0.0554492 , 0.08139372, 0.08254087, 0.08560054,
       0.05350252, 0.0534403 , 0.26331669, 0.09527704, 0.05861375,
       0.08728505, 0.06575169, 0.        , 0.04353225, 0.01871487,
       0.        , 0.05484345, 0.00531237, 0.07674541, 0.05085288,
       0.18422052, 0.        , 0.01208759, 0.0290784 , 0.04169477,
       0.12354122, 0.16140808, 0.        , 0.1301383 , 0.06445396,
       0.        , 0.        , 0.        , 0.10686088, 0.02734

In [269]:
similarity.shape

(707, 707)

every book has 707 value ,it calculate all distance , if distance is 1 then it show itself

In [270]:
np.where(data.index=='Year of Wonders')

(array([702], dtype=int64),)

In [271]:
np.where(data.index=='Year of Wonders')[0][0]

702

In [272]:
list(enumerate(similarity[0]))

[(0, 0.9999999999999998),
 (1, 0.09991370443681998),
 (2, 0.011894676421575648),
 (3, 0.0),
 (4, 0.0522923428754616),
 (5, 0.027035592129264218),
 (6, 0.08005246987594641),
 (7, 0.13379800079660079),
 (8, 0.03177829126649559),
 (9, 0.03573297811768626),
 (10, 0.022627095695491876),
 (11, 0.06592521898256122),
 (12, 0.02030399594727408),
 (13, 0.09425025422049896),
 (14, 0.13044639995428942),
 (15, 0.08089640554206895),
 (16, 0.10866787678155022),
 (17, 0.049692801008612855),
 (18, 0.024530524976685255),
 (19, 0.11405414701564792),
 (20, 0.0),
 (21, 0.13965274219894008),
 (22, 0.07645775981397283),
 (23, 0.05992324616320607),
 (24, 0.08499677059744355),
 (25, 0.0),
 (26, 0.06829593609316079),
 (27, 0.13307519236960055),
 (28, 0.0740492482078533),
 (29, 0.11854319325391062),
 (30, 0.007482996232171423),
 (31, 0.01435345067390976),
 (32, 0.0),
 (33, 0.07761014992531964),
 (34, 0.04406341116244771),
 (35, 0.015162595382829198),
 (36, 0.09251799313259884),
 (37, 0.017761989688713762),
 (38,

In [273]:
x=list(enumerate(similarity[0]))

In [274]:
list(enumerate(similarity[0]))[0][1]

0.9999999999999998

In [275]:
sorted(list(enumerate(similarity[0])),key=lambda x: x[1],reverse=True)[1:5]

[(47, 0.26331668558969296),
 (546, 0.25717553496685785),
 (82, 0.2306084004654351),
 (635, 0.22700367288326906)]

In [276]:
p=sorted(list(enumerate(similarity[0])),key=lambda x: x[1],reverse=True)[1:5]
for i in p:
    x=i[0]
    print(x)

47
546
82
635


In [277]:
data.index

Index(['1984', '1st to Die: A Novel', '2nd Chance', '4 Blondes',
       'A Bend in the Road', 'A Case of Need',
       'A Child Called \It\": One Child's Courage to Survive"',
       'A Civil Action', 'A Day Late and a Dollar Short', 'A Fine Balance',
       ...
       'Winter Solstice', 'Wish You Well', 'Without Remorse',
       'Wizard and Glass (The Dark Tower, Book 4)', 'Wuthering Heights',
       'Year of Wonders', 'You Belong To Me',
       'Zen and the Art of Motorcycle Maintenance: An Inquiry into Values',
       'Zoya', '\O\" Is for Outlaw"'],
      dtype='object', name='Book-Title', length=707)

In [278]:
data.index[78]

'Blood Work'

In [279]:
def recomand(book_name):
  index=np.where(data.index==book_name)[0][0]
  similar=sorted(list(enumerate(similarity[index])),key=lambda x:x[1],reverse=True)[1:6]
  
  for i in similar:
    temp=book[book['Book-Title']==data.index[i[0]]]
    return temp

In [280]:
recomand('1984')

ISBN   Book-Title    Book-Author Year-Of-Publication  \
285     0451526341  Animal Farm  George Orwell                2004   
4878    0451522303  Animal Farm  George Orwell                1956   
5912    0451524667  Animal Farm  George Orwell                1990   
8985    0736605673  Animal Farm  George Orwell                1981   
52635   0451518012  Animal Farm  George Orwell                1956   
53668   0451521560  Animal Farm  George Orwell                1986   
115681  0451519000  Animal Farm  George Orwell                1956   
116165  0151072558  Animal Farm  George Orwell                1990   
131735  0451520874  Animal Farm  George Orwell                1956   
209114  1854597892  Animal Farm  George Orwell                2004   

                             Publisher  \
285                             Signet   
4878                       Signet Book   
5912         New Amer Library Classics   
8985                     Books on Tape   
52635                      Signet Book   
53668   New Amer Library Classics (Mm)   
115681             Penguin Putnam~mass   
116165                        Harcourt   
131735                     Signet Book   
209114                 Nick Hern Books   

                                              Image-URL-S  \
285     http://images.amazon.com/images/P/0451526341.0...   
4878    http://images.amazon.com/images/P/0451522303.0...   
5912    http://images.amazon.com/images/P/0451524667.0...   
8985    http://images.amazon.com/images/P/0736605673.0...   
52635   http://images.amazon.com/images/P/0451518012.0...   
53668   http://images.amazon.com/images/P/0451521560.0...   
115681  http://images.amazon.com/images/P/0451519000.0...   
116165  http://images.amazon.com/images/P/0151072558.0...   
131735  http://images.amazon.com/images/P/0451520874.0...   
209114  http://images.amazon.com/images/P/1854597892.0...   

                                              Image-URL-M  \
285     http://images.amazon.com/images/P/0451526341.0...   
4878    http://images.amazon.com/images/P/0451522303.0...   
5912    http://images.amazon.com/images/P/0451524667.0...   
8985    http://images.amazon.com/images/P/0736605673.0...   
52635   http://images.amazon.com/images/P/0451518012.0...   
53668   http://images.amazon.com/images/P/0451521560.0...   
115681  http://images.amazon.com/images/P/0451519000.0...   
116165  http://images.amazon.com/images/P/0151072558.0...   
131735  http://images.amazon.com/images/P/0451520874.0...   
209114  http://images.amazon.com/images/P/1854597892.0...   

                                              Image-URL-L  
285     http://images.amazon.com/images/P/0451526341.0...  
4878    http://images.amazon.com/images/P/0451522303.0...  
5912    http://images.amazon.com/images/P/0451524667.0...  
8985    http://images.amazon.com/images/P/0736605673.0...  
52635   http://images.amazon.com/images/P/0451518012.0...  
53668   http://images.amazon.com/images/P/0451521560.0...  
115681  http://images.amazon.com/images/P/0451519000.0...  
116165  http://images.amazon.com/images/P/0151072558.0...  
131735  http://images.amazon.com/images/P/0451520874.0...  
209114  http://images.amazon.com/images/P/1854597892.0...

In [281]:
def recomand(book_name):
  index=np.where(data.index==book_name)[0][0]
  similar=sorted(list(enumerate(similarity[index])),key=lambda x:x[1],reverse=True)[1:5]
  d=[]
  for i in similar:
    item=[]
    temp=book[book['Book-Title']==data.index[i[0]]]
    item.extend(temp.drop_duplicates('Book-Title')['Book-Author'].values)
    item.extend(temp.drop_duplicates('Book-Title')['Book-Author'].values)
    item.extend(temp.drop_duplicates('Book-Title')['Image-URL-M'].values)
    d.append(item)
  return d
    

In [282]:
recomand('1984')

[['George Orwell',
  'George Orwell',
  'http://images.amazon.com/images/P/0451526341.01.MZZZZZZZ.jpg'],
 ['Margaret Atwood',
  'Margaret Atwood',
  'http://images.amazon.com/images/P/0449212602.01.MZZZZZZZ.jpg'],
 ['Aldous Huxley',
  'Aldous Huxley',
  'http://images.amazon.com/images/P/0060809833.01.MZZZZZZZ.jpg'],
 ['ANNE RICE',
  'ANNE RICE',
  'http://images.amazon.com/images/P/0345313860.01.MZZZZZZZ.jpg']]

In [283]:
def recomand(book_name):
  index=np.where(data.index==book_name)[0][0]
  similar=sorted(list(enumerate(similarity[index])),key=lambda x:x[1],reverse=True)[1:6]
  return similar

In [284]:
recomand('1984')

[(47, 0.26331668558969296),
 (546, 0.25717553496685785),
 (82, 0.2306084004654351),
 (635, 0.22700367288326906),
 (552, 0.2204467783292418)]

In [285]:
def recomand(book_name):
  index=np.where(data.index==book_name)[0][0]
  similar=sorted(list(enumerate(similarity[index])),key=lambda x:x[1],reverse=True)[1:6]
  res=[]
  for i in similar:
    item=[]
    temp=book[book['Book-Title']==data.index[i[0]]]
    item.extend(temp.drop_duplicates('Book-Title')['Book-Title'].values)
    item.extend(temp.drop_duplicates('Book-Title')['Book-Author'].values)
    item.extend(temp.drop_duplicates('Book-Title')['Image-URL-M'].values)
    res.append(item)
  return res

In [286]:
recomand('1984')

[['Animal Farm',
  'George Orwell',
  'http://images.amazon.com/images/P/0451526341.01.MZZZZZZZ.jpg'],
 ["The Handmaid's Tale",
  'Margaret Atwood',
  'http://images.amazon.com/images/P/0449212602.01.MZZZZZZZ.jpg'],
 ['Brave New World',
  'Aldous Huxley',
  'http://images.amazon.com/images/P/0060809833.01.MZZZZZZZ.jpg'],
 ['The Vampire Lestat (Vampire Chronicles, Book II)',
  'ANNE RICE',
  'http://images.amazon.com/images/P/0345313860.01.MZZZZZZZ.jpg'],
 ['The Hours : A Novel',
  'Michael Cunningham',
  'http://images.amazon.com/images/P/0312243022.01.MZZZZZZZ.jpg']]

In [287]:
import pickle

In [288]:
def recomand(book_name):
  index=np.where(data.index==book_name)[0][0]
  similar=sorted(list(enumerate(similarity[index])),key=lambda x:x[1],reverse=True)[1:6]
  res=[]
  for i in similar:
    item=[]
    temp=book[book['Book-Title']==data.index[i[0]]]
    item.extend(temp.drop_duplicates('Book-Title')['Book-Title'].values)
    item.extend(temp.drop_duplicates('Book-Title')['Book-Author'].values)
    item.extend(temp.drop_duplicates('Book-Title')['Image-URL-M'].values)
    res.append(item)
  return res

In [289]:
recomand('1984')

[['Animal Farm',
  'George Orwell',
  'http://images.amazon.com/images/P/0451526341.01.MZZZZZZZ.jpg'],
 ["The Handmaid's Tale",
  'Margaret Atwood',
  'http://images.amazon.com/images/P/0449212602.01.MZZZZZZZ.jpg'],
 ['Brave New World',
  'Aldous Huxley',
  'http://images.amazon.com/images/P/0060809833.01.MZZZZZZZ.jpg'],
 ['The Vampire Lestat (Vampire Chronicles, Book II)',
  'ANNE RICE',
  'http://images.amazon.com/images/P/0345313860.01.MZZZZZZZ.jpg'],
 ['The Hours : A Novel',
  'Michael Cunningham',
  'http://images.amazon.com/images/P/0312243022.01.MZZZZZZZ.jpg']]

In [290]:

def form1(x):
         
         index=np.where(data.index==x)[0][0]
         similar=sorted(list(enumerate(similarity[index])),key=lambda x:x[1],reverse=True)[1:5]
         res=[]
         for i in similar:
           item=[]
           temp=book[book['Book-Title']==data.index[i[0]]]
           item.extend(temp.drop_duplicates('Book-Title')['Book-Title'].values)
           item.extend(temp.drop_duplicates('Book-Title')['Book-Author'].values)
           item.extend(temp.drop_duplicates('Book-Title')['Image-URL-M'].values)
           res.append(item)
         return res

In [291]:
form1('1984')

[['Animal Farm',
  'George Orwell',
  'http://images.amazon.com/images/P/0451526341.01.MZZZZZZZ.jpg'],
 ["The Handmaid's Tale",
  'Margaret Atwood',
  'http://images.amazon.com/images/P/0449212602.01.MZZZZZZZ.jpg'],
 ['Brave New World',
  'Aldous Huxley',
  'http://images.amazon.com/images/P/0060809833.01.MZZZZZZZ.jpg'],
 ['The Vampire Lestat (Vampire Chronicles, Book II)',
  'ANNE RICE',
  'http://images.amazon.com/images/P/0345313860.01.MZZZZZZZ.jpg']]

In [292]:
final_data=pd.read_csv(r'C:\Users\deyar\OneDrive\Desktop\BookRecommendationSystem\artifacts\data.csv')

In [293]:
final_data

254  2276  2766  2977  3363  4017  4385  6251  6323  6543  ...  271705  \
0    9.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...    10.0   
1    0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   9.0  ...     0.0   
2    0.0  10.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...     0.0   
3    0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...     0.0   
4    0.0   0.0   7.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...     0.0   
..   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...     ...   
702  0.0   0.0   0.0   7.0   0.0   0.0   0.0   0.0   0.0   0.0  ...     0.0   
703  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...     0.0   
704  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...     0.0   
705  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...     0.0   
706  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...     0.0   

     273979  274004  274061  274301  274308  275970  277427  277639  278418  
0       0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
1       0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
2       0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
3       0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
4       0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
..      ...     ...     ...     ...     ...     ...     ...     ...     ...  
702     9.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
703     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
704     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
705     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
706     0.0     0.0     0.0     8.0     0.0     0.0     0.0     0.0     0.0  

[707 rows x 815 columns]